source/common/network/lc_trie.h

```diff
@@ -0,0 +1,540 @@
+      }
+
+      // Compute the number of bits required for branching by checking for all
+      // patterns (b=2 {00, 01, 10, 11}; b=3 {000,001,010,011,100,101,110,111}, etc)
+      // are covered in the set.
+      uint32_t branch = 1;
+      uint32_t count;
+      do {
+        ++branch;
+
+        // Check if the current branch factor with the fill factor can contain all of the nodes
+        // in the current range or if the current branching factor is larger than the
+        // IP address_size.
+        if (n < fill_factor_ * (1 << branch) ||
+            static_cast<uint32_t>(compute.prefix_ + branch) > address_size) {
+          break;
+        }
+
+        // Start by checking the bit patterns at ip_prefixes_[first] through
+        // ip_prefixes_[first + n-1].
+        index = first;
+        // Pattern to search for.
+        uint32_t pattern = 0;
+        // Number of patterns found while looping through the list.
+        count = 0;
+
+        // Search for all patterns(values) within 1<<branch.
+        while (pattern < static_cast<uint32_t>(1 << branch)) {
+          bool pattern_found = false;
+          // Keep on looping until either all nodes in the range have been visited or
+          // an IP prefix doesn't match the pattern.
+          while (index < first + n &&
+                 static_cast<uint32_t>(extractBits<IpType, address_size>(
+                     compute.prefix_, branch, ip_prefixes_[index].ip_)) == pattern) {
+            ++index;
+            pattern_found = true;
+          }
+
+          if (pattern_found) {
+            ++count;
+          }
+          ++pattern;
+        }
+        // Stop iterating once the size of the branch (with the fill factor ratio) is able to
+        // contain all of the prefixes within the current range of ip_prefixes_[first] to
+        // ip_prefixes_[first+n-1].
+      } while (count >= fill_factor_ * (1 << branch));
+
+      // Set the branching factor.
+      compute.branch_ = branch - 1;
+      return compute;
+    }
+
+    /**
+     * Recursively build a trie for IP prefixes from position 'first' to 'first+n-1'.
+     * @param prefix supplies the prefix to ignore when building the sub-trie.
+     * @param first supplies the ip_prefixes_ index into for this sub-trie.
+     * @param n suppplies the number of entries for the sub-trie.
+     * @param position supplies the root for this sub-trie.
+     * @param next_free_index supplies the next available index in the trie_.
+     */
+    void buildRecursive(uint32_t prefix, uint32_t first, uint32_t n, uint32_t position,
+                        uint32_t& next_free_index) {
+      // Setting a leaf, the branch and skip are 0.
+      if (n == 1) {
+        trie_[position].address_ = first;
+        return;
+      }
+
+      ComputePair output = computeBranch(prefix, first, n);
+
+      uint32_t address = next_free_index;
+      trie_[position].branch_ = output.branch_;
+      trie_[position].skip_ = output.prefix_ - prefix;
```

Comment:
Synced with Jose offline - i'll add  a comment that the skip value is the number of bits between the last prefix at prefix and the output.prefix . (Mention that it is the number of bits)